### PR TITLE
Pubmatic: Fix Shared Memory Overwriting

### PR DESCRIPTION
**Pull request by guscarreon**
Copy of reference field Banner needed to be made inside `adapters/pubmatic/pubmatic.go` in order to avoid a data race on shared memory. Authors of source code should be notified and encourage to make any changes to this PR if they disagree with the proposed approach.

### CODE REVIEW

**Comment by SyntaxNode** [comment]
Please contact PubMatic using their bidder info email address to inform them of this change.

File: adapters/pubmatic/pubmatic.go [hunk 1]
```diff
@@ -158,8 +158,7 @@ func (a *PubmaticAdapter) Call(ctx context.Context, req *pbs.PBSRequest, bidder
 					}
 
 					pbReq.Imp[i].TagID = strings.TrimSpace(adSlot[0])
-					pbReq.Imp[i].Banner.H = openrtb.Uint64Ptr(uint64(height))
-					pbReq.Imp[i].Banner.W = openrtb.Uint64Ptr(uint64(width))
+					pbReq.Imp[i].Banner = assignBannerHeightAndWidth(pbReq.Imp[i].Banner, uint64(height), uint64(width))
```
Comment (SyntaxNode) [inline adapters/pubmatic/pubmatic.go:161]: Don't bother with changes in the old interfaces, such as `Call`. That shouldn't be caught by the json test framework.

File: adapters/pubmatic/pubmatic.go [hunk 2]
```diff
@@ -428,25 +426,27 @@ func validateAdSlot(adslot string, imp *openrtb.Imp) error {
...
 }
 
-func assignBannerSize(banner *openrtb.Banner) error {
-	if banner == nil {
-		return nil
-	}
+func assignBannerHeightAndWidth(banner *openrtb.Banner, h uint64, w uint64) *openrtb.Banner {
+	bannerCopy := *banner
+
+	bannerCopy.W = new(uint64)
```
Comment (SyntaxNode) [inline adapters/pubmatic/pubmatic.go:432]: Please use the `openrtb.Uint64Ptr()` helper method. It's a bit cleaner:
```
bannerCopy.W = openrtb.Uint64Ptr(w)
bannerCopy.H = openrtb.Uint64Ptr(h)
```

**Comment by guscarreon** [comment]
@sachin-pubmatic,
Recently coded pull request #1756 is meant to enable Prebid Server to find data race conditions in adapter bidder source code. According to this newly coded feature, your adapter might be writing to shared memory and this PR implements an approach to prevent it. Please let us know whether or not you agree with the changes proposed in this pull request or if you rather want to code the fix yourself. 

When it comes to adapters' source code improvements, we usually wait a week for owners to come back to us. If we get no feedback, we will go ahead to review and merge this pull request ourselves.

We'd love to hear your take on this issue.

File: adapters/pubmatic/pubmatic.go (hunk 1 above)
Comment (guscarreon) [reply adapters/pubmatic/pubmatic.go:161]: Rolled back

File: adapters/pubmatic/pubmatic.go [hunk 3]
```diff
@@ -428,25 +427,25 @@ func validateAdSlot(adslot string, imp *openrtb.Imp) error {
...
-
+func assignBannerSize(banner *openrtb.Banner) (*openrtb.Banner, error) {
 	if banner.W != nil && banner.H != nil {
-		return nil
+		return banner, nil
 	}
 
 	if len(banner.Format) == 0 {
-		return errors.New(fmt.Sprintf("No sizes provided for Banner %v", banner.Format))
+		return banner, errors.New(fmt.Sprintf("No sizes provided for Banner %v", banner.Format))
```
Comment (SyntaxNode) [inline adapters/pubmatic/pubmatic.go:436]: I think this should return a `nil` banner pointer if there's an error.

Comment (guscarreon) [reply adapters/pubmatic/pubmatic.go:436]: Good catch. Removed

**Review by hhhjort — APPROVED** [review]
LGTM

**Review by SyntaxNode — APPROVED** [review]

**Comment by SyntaxNode** [comment]
We received approval via email for this change.

**Comment by bjorn-lw** [comment]
I just downloaded and tested the latest version of Prebid server (0.155) and noticed that with this version Pubmatic is starting to respond with unexpected banner sizes.

**Comment by bjorn-lw** [comment]
I just downloaded and tested the latest version of Prebid server (0.155) and noticed that with that version Pubmatic is starting to respond with unexpected banner sizes.

For instance, a request that requests only 160x600 as size, Pubmatic may respond with 600x600.

If I use an earlier version of Prebid Server (sorry, I can currently only test 0.139.1 easily) this problem does not occur.

But after looking at what has changed in the pubmatic adapter, this looks like something that maybe is causing this. I have no idea why though, and I may be completely off.

I will compile earlier versions and test until the problem goes away to see exactly in which version it happens.

**Comment by bjorn-lw** [comment]
The issue is NOT present in 0.152.0 so this is most likely related to this change.

Sorry for not being too helpful in locating the source of the issue, GO is not my first language, but I will make an honest try.

**Pull request merged**